File: bdp-devops/ambari/ambari-server/src/main/python/ambari_server/ambariPath.py
#!/usr/bin/env python
"""
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Ambari Agent

"""

import os
import re

AMBARI_SERVER_ROOT_ENV_VARIABLE = "ROOT"

class AmbariPath():
  root_directory = os.getenv(AMBARI_SERVER_ROOT_ENV_VARIABLE, "/")

  @staticmethod
  def get(path):
    """
    Any paths which to ambari-server files, should be wrapped with this function call.
    Which is needed for the situations when ambari-server installed not in / but in other folder like /opt.
    Not ambari paths like /var/run/postgresql SHOULD NOT wrapped by this call though.
    """
    #return os.path.realpath(AmbariPath.root_directory + os.sep + path) # realpath won't replace slashes for python2.6
    return re.sub('/+', '/', AmbariPath.root_directory + os.sep + path)
    
  

File: bdp-devops/ambari/ambari-server/src/main/java/org/apache/ambari/server/utils/AmbariPath.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ambari.server.utils;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class AmbariPath {
  private static final Logger LOG = LoggerFactory.getLogger(AmbariPath.class);
  
  public static final String AMBARI_SERVER_ROOT_ENV_VARIABLE = "ROOT";
  public static final String rootDirectory = System.getenv(AMBARI_SERVER_ROOT_ENV_VARIABLE);
  
  public static String getPath(String path) {
    if(rootDirectory == null) {  
      LOG.warn("Cannot get $ROOT enviroment variable. Installed to custom root directory Ambari might not work correctly.");
      return path;
    }
    String result = (rootDirectory + path).replaceAll("/+","/");
    return result;
  }
}
